### PR TITLE
Changes to support ADFS/WS-Federation SAML2.0

### PR DESCRIPTION
**Pull request by sandeep048**
Added few extra settings to play nice with Microsoft ADFS/WS-Federation SAML2.0 server.

### CODE REVIEW

File: src/onelogin/saml2/authn_request.py [hunk 1]
```diff
@@ -47,9 +47,15 @@ def __init__(self, settings, force_authn=False, is_passive=False):
...
-            name_id_policy_format = OneLogin_Saml2_Constants.NAMEID_ENCRYPTED
+        name_id_policy_format = None
+        if 'wantNameId' in security and security['wantNameId']:
+            name_id_policy_format = sp_data['NameIDFormat']
+            if 'wantNameIdEncrypted' in security and security['wantNameIdEncrypted']:
+                name_id_policy_format = OneLogin_Saml2_Constants.NAMEID_ENCRYPTED
+
+        name_id_policy = ''
+        if name_id_policy_format: 
+            name_id_policy = """<samlp:NameIDPolicy Format="%(name_id_policy)s" AllowCreate="true" />""" % name_id_policy_format
```
Comment (pitbulk) [inline src/onelogin/saml2/authn_request.py]: Why instead don't set at SP settings, the NameIDFormat as "urn:oasis:names:tc:SAML:2.0:nameid-format:unspecified"

File: src/onelogin/saml2/authn_request.py [hunk 2]
```diff
@@ -77,9 +83,11 @@ def __init__(self, settings, force_authn=False, is_passive=False):
                 authn_comparison = security['requestedAuthnContextComparison']
 
             if security['requestedAuthnContext'] is True:
-                requested_authn_context_str = """    <samlp:RequestedAuthnContext Comparison="%s">
-        <saml:AuthnContextClassRef>urn:oasis:names:tc:SAML:2.0:ac:classes:PasswordProtectedTransport</saml:AuthnContextClassRef>
-    </samlp:RequestedAuthnContext>""" % authn_comparison
+                if 'activeDirectoryFederation' in security and security['activeDirectoryFederation']:
```
Comment (pitbulk) [inline src/onelogin/saml2/authn_request.py]: Don't need that.

Instead set as security settings, the value 'requestedAuthnContext' as urn:federation:authentication:windows
as described in the [documentation](https://github.com/onelogin/python-saml#settings)

File: src/onelogin/saml2/response.py [hunk 3]
```diff
@@ -128,14 +128,15 @@ def is_valid(self, request_data, request_id=None):
...
                 # Checks destination
-                destination = self.document.get('Destination', '')
-                if destination:
-                    if not destination.startswith(current_url):
-                        # TODO: Review if following lines are required, since we can control the
-                        # request_data
-                        #  current_url_routed = OneLogin_Saml2_Utils.get_self_routed_url_no_query(request_data)
-                        #  if not destination.startswith(current_url_routed):
-                        raise Exception('The response was received at %s instead of %s' % (current_url, destination))
+                if security.get('checkDestination', True):
```
Comment (pitbulk) [inline src/onelogin/saml2/response.py]: Check the destination is a MUST on the SAML standard so I can't make it optional.

SAML Bindings spec contains the following under security considerations for Redirect and POST bindings:
```
If the message is signed, the Destination XML attribute in the root SAML element of the protocol message MUST contain the URL to which the sender has instructed the user agent to deliver the message. The recipient MUST then verify that the value matches the location at which the message has been received.
```

**Comment by pitbulk** [comment]
I must reject your PR, you don't need those changes in order to be compatible with ADFS/WS-Federation.

Read my comments.

File: src/onelogin/saml2/authn_request.py (hunk 1 above)
Comment (sandeep048) [inline src/onelogin/saml2/authn_request.py]: nameid-format:unspecified is not working. For our specific case we wouldn't want the <samlp:NameIDPolicy> tag at all in the generated SAML.

File: src/onelogin/saml2/authn_request.py (hunk 2 above)
Comment (sandeep048) [inline src/onelogin/saml2/authn_request.py]: If so then we will lose "xmlns:samlp="urn:oasis:names:tc:SAML:2.0:assertion" tag from saml:AuthnContextClassRef and the generated SAML markup.

File: src/onelogin/saml2/response.py (hunk 3 above)
Comment (sandeep048) [inline src/onelogin/saml2/response.py]: Destination match should be relaxeable if domain is a correct match, this if for testing purposes.

File: src/onelogin/saml2/authn_request.py (hunk 2 above)
Comment (sandeep048) [inline src/onelogin/saml2/authn_request.py]:
```
<samlp:RequestedAuthnContext Comparison="exact">
  <saml:AuthnContextClassRef xmlns:samlp="urn:oasis:names:tc:SAML:2.0:assertion">urn:federation:authentication:windows</saml:AuthnContextClassRef>
</samlp:RequestedAuthnContext>

```

File: src/onelogin/saml2/response.py (hunk 3 above)
Comment (pitbulk) [inline src/onelogin/saml2/response.py]: Sorry, the specs says that "MUST contain the URL", not only the domain 
Why are you not able to set the right value on the IdP?

File: src/onelogin/saml2/authn_request.py (hunk 2 above)
Comment (pitbulk) [inline src/onelogin/saml2/authn_request.py]: you don't need to define the samlp namespace there, is already set at the AuthNRequest element.

File: src/onelogin/saml2/authn_request.py (hunk 1 above)
Comment (pitbulk) [inline src/onelogin/saml2/authn_request.py]: Ok, this is something that I will consider in a future release, to be able to not define NameIDPolicy element.

**Pull request closed without merging**